File: codes/3.ProgrammingLanguage/编程语言试水/Score100WordsAnalyzer.cpp
#include <chrono>
#include <fstream>
#include <iostream>
#include <sstream>
#include <vector>

int main() {
  auto start = std::chrono::high_resolution_clock::now();
  std::ifstream inputFile("./Dictionary.txt");
  std::string line;
  std::vector<std::string> score100Words;
  while (getline(inputFile, line)) {
    if (line.contains(' ') || line.contains('-') || line.contains('=')) {
      continue;
    }
    std::istringstream is(line);
    std::string word;
    is >> word;
    int score = [](const std::string &word) {
      int score = 0;
      for (auto &c: word) {
        score += (c - 'a' + 1);
      }
      return score;
    }(word);
    if (score == 100) {
      score100Words.push_back(word);
    }
  }
  for (auto &word: score100Words) {
    std::cout << word << std::endl;
  }
  auto end = std::chrono::high_resolution_clock::now();
  auto time_diff = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
  std::cout << "\nwords number: " << score100Words.size() << std::endl;
  std::cout << "        time: " << time_diff.count() << "ms" << std::endl;
  return 0;
}

/*
 * 编码体验：
 * 1、没有链式函数式编程（也可能是我不会写）
 * 2、IDE（CLion）中点击方法看详情时，看到的是头文件中的定义，其中又看不到对应的文档，比如方法含义、参数含义、返回值含义等
 */

File: codes/3.ProgrammingLanguage/编程语言试水/Score100WordsAnalyzer.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"time"
)

func main() {
	start := time.Now()
	file, _ := os.OpenFile("Dictionary.txt", os.O_RDONLY, 0666)
	defer file.Close()
	score100Words := make([]string, 0)
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		text := scanner.Text()
		if !strings.ContainsAny(text, " -=") {
			score := 0
			for _, c := range text {
				score += int(c) - 'a' + 1
			}
			if score == 100 {
				score100Words = append(score100Words, text)
			}
		}
	}
	fmt.Println(score100Words)
	fmt.Println("\nwords number: ", len(score100Words))
	fmt.Println("        time: ", time.Since(start).Milliseconds(), "ms")
}

/**
编码体验：
1、没有分号
2、方法可以返回多个值，_ 表示忽略
3、defer 关键字
4、strings.ContainsAny 在其他语言中没有这个方法，但是很实用
5、未使用的变量会直接报错
6、本例中并没有类似Java、C#、F#那样的链式编程，反而更像C++面向过程的写法
*/
